src/core/federation.ts

```diff
@@ -281,6 +287,123 @@ function isANodeJSCodedError(e: Error): e is NodeJSCodedError {
   return typeof (e as NodeJSCodedError).code === "string";
 }
 
+/**
+ * A network sender is a reactor containing a portAbsentReaction.
+ */
+export class NetworkSender extends Reactor {
+  /**
+   * The last reaction of a NetworkSender reactor is 'portAbsentReactor'.
+   * @returns portAbsentReactor of this NetworkSender reactor
+   */
+  public getLastReactioOrMutation(): Reaction<Variable[]> | undefined {
+    return this._getLastReactionOrMutation();
+  }
+}
+
+/**
+ * A network receiver is a reactor handling a network input.
+ */
+export class NetworkReceiver<T> extends Reactor {
+  /*
+   * A FederatePortAction instance of this NetworkReceiver.
+   */
+  private networkInputAction: FederatePortAction<T> | undefined = undefined;
+
+  // The port ID of networkInputAction.
+  private readonly portID: number;
+
+  constructor(parent: Reactor, portID: number) {
+    super(parent);
+    this.portID = portID;
+  }
+
+  /**
+   * Getter for portID of this NetworkReactor.
+   */
+  public getPortID(): number | undefined {
+    return this.portID;
+  }
+
+  /**
+   * Register a federate port's action with the network receiver.
+   * @param networkInputAction The federate port's action for registration.
+   */
+  public registerNetworkInputAction(
+    networkInputAction: FederatePortAction<T>
+  ): void {
+    this.networkInputAction = networkInputAction;
+  }
+
+  /**
+   * Handle a message being received from the RTI.
+   * @param portID The destination port ID of the message.
+   * @param value The payload of the message.
+   */
+  public handleMessage(portID: number, value: T): void {
+    // Schedule this federate port's action.
+    // This message is untimed, so schedule it immediately.
+    if (portID !== this.portID) {
+      this.util.reportError(
+        "FederatedApp attempts to pass the tagged message to the wrong port ID"
+      );
+      return;
+    }
+    if (this.networkInputAction !== undefined) {
+      this.networkInputAction
+        .asSchedulable(this._getKey(this.networkInputAction))
```

Comment:
Since `networkInputAction` is `private`, why don't we store the schedulable action instead of retrieving it every time we schedule?

Reply:
This is solved in the commit [825f8fe](https://github.com/lf-lang/reactor-ts/pull/186/commits/825f8fe3ea973c185be40de458583d3dcb00862b)

Reply:
Addressed in https://github.com/lf-lang/reactor-ts/pull/186/commits/825f8fe3ea973c185be40de458583d3dcb00862b.

Reply:
You beat me to it :rofl:

Reply:
I won! :laughing: